### ISSUE TITLE
Add pretrained checkpoints for Bert

### ISSUE THREAD

**Issue by jbischof** [keras-team/keras-nlp#297]
# Problem
We exposed an API for fully configurable and fixed Bert architectures in PR #288. Most users will not have the resources to do pretraining on their own, however. We should offer checkpoints for `models.BertBase` similar to those used in the original paper.

# Implementation

## Weights

Model checkpoints from the original paper repo ([gh](https://github.com/google-research/bert)) or similar source can be converted to be compatible with our implementation in PR #288. This code will be a one-off task and not added to the repo.

## Format

`keras-cv` is currently using `h5`, but that format is deprecated. An obvious alternative is a `tf.checkpoint` using `model.save_weights()`. The resulting `data` and `index` files will have to be zipped and uploaded to our cloud storage bucket.

## API

There some prior art in `keras-cv` and our pretrained vocab to use a string to identify a model of interest which is then checked against a dictionary (e.g., [weights.py](https://github.com/keras-team/keras-cv/blob/master/keras_cv/models/weights.py#L58)). However, this requires the user to dive into the code to find possible names and a lot of input validation and error handling in the implementation.

I propose to create an `enum` for each `Model` class (e.g., `models.BertBase`) with possible checkpoints that will expose these options as part of the API and work with users' autocomplete engines:
```
class BertBaseCkpt(Enum):
   uncased = 0
   cased = 1
   zh = 2
```

Under the hood this `enum` would then be the key for a dictionary of metadata for each model including checkpoint location, vocabulary location, and a description string explaining where the model came from and what data was used to train it. We could also write a method to print the description for each/all `enum` value(s) to again protect the user from having to read through utility functions.

```
base_models = {
   BertBaseCkpt.uncased: {
      "ckpt": ...
      "description": ...
      "vocab": ...
   }
}
```

Model checkpoints will be hosted in our Google Cloud storage bucket and loaded behind the scenes given an enum value. The user should never have to interact with this dictionary.

Here is the enhanced API:
```
def BertBase(ckpt=None, name=None, trainable=True):
    """Bi-directional Transformer-based encoder network (Bert) using "Base"
    architecture.

    This network implements a bi-directional Transformer-based encoder as
    described in ["BERT: Pre-training of Deep Bidirectional Transformers for
    Language Understanding"](https://arxiv.org/abs/1810.04805). It includes the
    embedding lookups and transformer layers, but not the masked language model
    or classification task networks.

    Args:
        ckpt: BertBaseCkpt. A checkpoint of pretrained weights to load into the model. 
             If `None`, the model will be randomly initialized.
        name: String, optional. Name of the model.
        trainable: Boolean, optional. If the model's variables should be
            trainable.
```

## Vocabulary
Note that this proposal does not cover the vocabulary file, which will be incorporated in our preprocessing work. One potential proposal is to host the `vocab.txt` file used to train the model in the same folder to be loaded if the user does not supply their own. The rationale is that except for very advanced use cases the vocabulary must match that used in training, and requiring the user to specify their own will expose our Cloud storage implementation to the user for little gain. However, this is beyond the scope of the proposal.

**Comment by chenmoneygithub** [comment]
Thanks for putting up the nice design!

Some thoughts from my side:
1. Enum is something not commonly used in Keras code. My take is in general we prefer to use string identifier, @fchollet Tagging Francois for thoughts.
2. The arg `ckpt` is vague to our users, also it's not the actual checkpoint path. HuggingFace use a static method `from_pretrained`, which has a clear meaning "it is loading from pretrained model, and the pretrained model is XXX". 
3. checkpoint format is more preferrable than h5 saving format, as the later is technically deprecated.

Additionally, I would like to include the design of tokenizer saving/loading inside this discussion. A pretrained model should always come with its tokenizer, so we may not want to decouple the design. HuggingFace decouples the tokenizer from model, but uses the same string identifer, e.g., "bert-base-uncased" to glue those together. It seems we want to have tokenizer exist as a member of model, we should get the agreement.

**Comment by mattdangerw** [comment]
Thanks for the proposal! Some quick thoughts, will dig in more next week.

 - Using tf checkpoints sgtm. But probably good to check with KerasCV folk here. Will they be using tf checkpoints instead of h5 for newer weights they are hosting?
 - Arg name should probably still be `weights`, unless we have a good reason to change. We want a consistent API experience across keras as much as possible.
 - Also unsure about the enum type. Probably makes sense to follow the prior art in Keras unless we have a strong reason.
 - I wouldn't totally through away the checkpoint conversion code. We should try to keep all these in a tools or scripts directory, outside of `keras_nlp/' (the dir with library code). We should be able to reproduce the conversion as needed in the future, using checked in code (though not shipped code in our library).

**Comment by jbischof** [comment]
Thanks for the comments! Talking to @LukeWood the model loading experience for `keras-cv` is not yet polished. The point of this proposal was to explore the design space so we can optimize the user experience across both packages. Maybe `keras-cv` will want to use our new method. Rejecting anything new out of hand means we cannot innovate.

**Comment by mattdangerw** [comment]
Yeah sorry didn't mean to reject out of hand! Agreed we should asses when and where we want to diverge from the existing keras application design patterns individually. But we should avoid doing this too much because it comes at a cost when looking at the keras ecosystem as a whole.

In this particular case, I think what you are proposing is against the keras API design style guide.

> When using enums, make sure that their values are strings, so as to make it possible for users to pass plain strings (example: data_format="channels_last", padding="valid").

https://github.com/keras-team/governance/blob/master/keras_api_design_guidelines.md#seek-to-minimize-cognitive-load-for-our-users

So we should either stick to the existing pattern for enums, or make a deliberate choice to change this guidance across keras as a whole.
